Document bugzilla integration (#115)
openshift/ci-docs: content/en/docs/architecture/bugzilla.md

---
title: "Bugzilla Integration"
description: An overview of the integration of Bugzilla with the Openshift CI
---

## Linking Pull Requests to Bugzilla bugs

Pull requests can reference a Bugzilla bug in their title, for example
`Bug 123: Rest of the PR title`. This makes the bugzilla plugin run a couple

> of checks:

Reviewer: people have occasionally asked me if it's possible to link a PR to multiple bugs using this mechanism.  I believe the answer is still no, but can you mention that explicitly somewhere here?

Author: Done

> * If a bug with the id `123` exists and is in state `open`

Reviewer: is this superceded by the "correct state" configuration below?

Author: No, `is_open` is a boolean bug attribute, my best guess is that a bug is open if it is in any state that is not `Closed`. The rest api reference isn't really very helpful on that either:
```
is_open | boolean | true if this bug is open, false if it is closed.
```

Ref:  https://bugzilla.readthedocs.io/en/latest/api/core/v1/bug.html#get-bug

Reviewer: weird.  So you can both set the "is_open: true" field *and* "valid_state: closed" ?

guessing is_open predates valid_state.  Or am I misunderstanding how valid_state is used?

Reviewer: Also should this instead say "if a bug with the id `123` exists and is in the required open or closed state as configured by is_open [link to example config]"  ?

since just like everything else, it's configurable on a per repo basis.  Maybe mention our default is `open` but ultimately you are straddling a line between documenting what the plugin allows to be configured, and how our repositories are configured (and how our repositories are configured by default vs overridden specifically).

Author: > weird. So you can both set the "is_open: true" field and "valid_state: closed" ?

Technically yes, but a bug will never be able to satisfy both requirements. It will however be able to satisfy something like `is_open: true` and `valid_state: New`

* If the target release of the bugzilla bug matches the configured target release for the branch the PR points at. This mapping is
	configured in the [openshift/release repository][0] via the `target_release` property

> * If the bug is in the correct state. Accepted states are configured in the [openshift/release repository][0] through the `valid_states`

Reviewer: i'm really happy to see you linked to an example, but i think people may miss this link because it's specifically tied to talking about the valid_states property.  I would include a direct mention of "for an example of how the plugin is configured for a specific repository, see:" and link to an example of that (not the default config) that illustrates the various config properties.

  property

> * If the bug has `dependant_bugs` and `dependent_bug_states` configured, those dependant bugs must be in one of the configured states

Reviewer: ```suggestion
* If the bug has `dependent_bugs` and `dependent_bug_states` configured, those dependent bugs must be in one of the configured states
```
also do you mean if the repository has dependent_bugs configured?

Author: Interesting, `dependant` is only correct in Britisch English, TILd

Reviewer: huh.  I didn't realize it was a british spelling.

If all of these checks succeed, the Bugzilla plugin will link the Pull Request in Bugzilla. If the configuration for the given
github org, branch or repository contains the `state_after_validation` setting, the Bugzilla bug will be moved to that state
after all validations succeeded.

Similarly, the `state_after_close` and `state_after_merge` setting will cause the Bugzilla bug to be moved into the referenced
state after close/merge.

The Bugzilla automation will always comment on GitHub and mention the validations it ran. If any of the validations had an error,
it will be there. Also, it adds `bugzilla/valid-bug` or `bugzilla/invalid-bug` label to the Pull Requests. Pull Requests with the
`bugzilla/invalid-bug` label are never allowed to merge. Release branches in OCP usually require the `bugzilla/valid-bug` label.

## Cherry Picking

The Bugzilla automation can also assist in cherry picking. After an automated cherry-pick was created for a Pull Request that
referenced a Bugzilla bug, the automation will clone the original bug with the new target release and retitle the PR to correctly

> reference that new Bug. The new bug will also be marked as dependant on the original bug.

Reviewer: ```suggestion
reference that new Bug. The new bug will also be marked as dependent on the original bug.
```

---
title: "Bugzilla Integration"
description: An overview of the integration of Bugzilla with the Openshift CI
---

Many automated processes exist to help you navigate the OpenShift Bugzilla workflow when you're fixing a Bugzilla bug with a
GitHub Pull Request. Read up on the overall process [here][6] before diving into the specifics of automation.

## Linking Pull Requests to Bugzilla bugs

In order to declare that your pull request will fix a Bugzilla bug, all you need to do is add the bug number as a prefix to your pull
request title, like `Bug 123: Rest of the PR title`. A PR can at most reference one Bugzilla bug.

This will make the integration run a couple of validations. It will comment on GitHub with a detailled explanation of the
validations it ran and their result, so you can easily fix issues if needed. It will also add a label that indicates the success
of those validations to integrate with the merge automation.

You can find a detailed list of the bugzilla plugin settings for your repo in the [Prow Plugins][5] page by first selecting your

> repository in the top left and then clicking on `details` of the Bugzilla plugin.

Reviewer 2: nit: we should note to use `/bugzilla refresh` if Bugzilla state changes and re-evaluation is needed

## Cherry Picking/Backporting

The Bugzilla automation can also assist in the backporting process by integrating with the cherry picking automation. After an
automated cherry-pick was created for a Pull Request that referenced a Bugzilla bug, the automation will clone the original bug

> with the new target release and retitle the PR to correctly reference that new Bug. The new bug will also be marked as dependent on the original bug.

Reviewer 2: nit: we should note that a cherry-pick can be done with `/cherry-pick <branch>` and will happen after a PR merges

Reviewer 2: nit: we should note today that you have to cherrypick A to B, then B to C, in order for the bugs to get correctly chained (even though you can technically do A to C at the same time)

Reviewer 2: nit: we should mention using bugs.ci.openshift.org when you want to declare that a bug *should* be fixed in other versions but don't have the PR(s) ready to backport yet

---
title: "Bugzilla Integration"
description: An overview of the integration of Bugzilla with the Openshift CI
---

> **Note:** This documentation describes how to configure CI to assist with Bugzilla-related processes. It does _not_ describe

Reviewer 2: Again, I really don't think this is the (only) doc we want or need. The point of this doc is to alleviate questions from help-desk. Questions do not come into helpdesk about "how does `valid_states` get used to validate my PR?".

I think we should focus on writing a doc that fulfills this user story from the Jira card:

As a author of a PR to an OpenShift component
I want to understand how my PR may interact with Bugzilla, what the normal workflows are and what I need to keep in mind
So that I can be effective at fixing bugs with pull requests

So, as an example, instead of this preamble, I'd perhaps write a sentence like:

> Many automated processes exist to help you navigate the OpenShift Bugzilla workflow when you're fixing a Bugzilla bug with a GitHub Pull Request. Read up on the overall process [here](...) before diving into the specifics of automation.

Reviewer 3: Is this a call to inline the current backport-process Google Doc?  Because I'm in favor of Git over Google Docs for version control, but I'm not sure if everyone else is on board with that or not :p.

Reviewer 3: Never mind, I see the linked Source page [linking to Google Docs in turn][1].

[1]: https://source.redhat.com/groups/public/atomicopenshift/atomicopenshift_wiki/openshift_bugzilla_process#jive_content_id_Fixing_in_multiple_releases

how these processes look. Documentation about the Openshift Bugzilla process can be found [here][6].

## Linking Pull Requests to Bugzilla bugs

> Pull requests can reference a Bugzilla bug in their title, for example

Reviewer 2: Similarly, when we consider the target audience this type of sentence changes:

> In order to declare that your pull request will fix a Bugzilla bug, all you need to do is add the bug number as a prefix to your pull request title, like `Bug 123: Rest of the PR title`. A PR can at most reference ...

`Bug 123: Rest of the PR title`. A PR can at most reference one Bugzilla bug. Referencing a bug makes the bugzilla plugin run a couple
of checks:

> * If a bug with the id `123` exists

Reviewer 2: I think for the general overview, we should let the plugin do it's thing and not over-explain here. Think about - for any of these validations - is the user who's fixing the bug responsible for _doing anything_ themselves? If they've been assigned a bug to work on, I don't think they have to do anything necessarily to make it valid, and if so the plugin will let them know. So, from an instructional point-of-view when we're describing what the user has to be doing and thinking about, we don't need this level of detail.

Reviewer 3: What sorts of things does the plugin not explain in-PR?  [Here][1] is the bot explaining that I need a `Bug ###` reference.  Once I add one, it will comment explaining the verification checks and which, if any, have failed.  How many help-desk threads are "I am not reading those in-PR comments", how many are "the in-PR comments are not answering my questions", and how many are "the in-PR comments could not possibly answer my questions"?  It's not clear to me if anything falls into that third category, and it's not clear to me that the first category can be resolved by adding more docs further away from the PR at issue.

[1]: https://github.com/openshift/cluster-version-operator/pull/536#issuecomment-806003890

Reviewer 2: I agree, @wking. That's why I think what we need here is a minimal "in order to fix a bug with your PR, add the prefix to your title" message and let the plugin run its course after that.

* If repo has the [`is_open: true`][3] setting, the bug must not be closed
* If the target release of the bugzilla bug matches the configured target release for the branch the PR points at. This mapping is

>   configured in the openshift/release repository via the `target_release` property. [Here][2] for example the `release-4.6` branch of the

Reviewer 2: However, having a section at the bottom of the doc with a reference snippet of configuration (or two) that show off how to configure the plugin and use effective overrides may be helpful. These details belong there.

After a PR that references a Bugzilla bug merges, the bug will be put into the state that is configured as [`state_after_merge`][4].
This is commonly used to trigger QE verification.

You can find a detailed list of the bugzilla plugin settings for your repo in the [Prow Plugins][5] page by first selecting your
repository in the top left and then clicking on `details` of the Bugzilla plugin.

## Cherry Picking/Backporting

The Bugzilla automation can also assist in the backporting process by integrating with the cherry picking automation. After an
automated cherry-pick was created for a Pull Request that referenced a Bugzilla bug, the automation will clone the original bug
with the new target release and retitle the PR to correctly reference that new Bug. The new bug will also be marked as dependent on the original bug.

## Integration with QE verification

Changes in OCP always end up in a nightly release assembled by the release-controller. The release-controller will run
a set of verification jobs and if those pass, the release is considered accepted. Once that happens, the release-controller
will:

> * Find all BZs associated with a change in the given release

Reviewer 2: This may be a nuanced point but I believe it's better here to write again as if we are speaking to the user. We don't necessarily need to have a reference manual for how the release controller works. I'd perhaps write something more like this:

> Normally, QE will verify that a fix correctly resolves a Bugzilla bug once the bug transitions to the `ON_QA` state, which happens after the PR merges and the `release-controller` has assembled a new release payload containing the code.

> However, if a QE engineer is assigned to the Bugzilla bug, they will also get notified when a pull request is created to fix the bug. The QE engineer can leave an approving review on the pull request to indicate that the fix has been verified before the PR even merges. In this case, when the nightly release containing the fix is accepted, the Bugzilla bug automatically transitions past the `ON_QE` state and into `VERIFIED`.

In order to declare that your pull request will fix a Bugzilla bug, all you need to do is add the bug number as a prefix to your pull
request title, like `Bug 123: Rest of the PR title`. A PR can at most reference one Bugzilla bug.

This will make the integration run a couple of validations. It will comment on GitHub with a detailed explanation of the
validations it ran and their result, so you can easily fix issues if needed. It will also add a label that indicates the success
of those validations to integrate with the merge automation.

You can always comment with `/bugzilla refresh` to re-run the validations. This is mostly useful if the Bugzilla bug was changed.

You can find a detailed list of the Bugzilla plugin settings for your repo in the [Prow Plugins][5] page by first selecting your
repository in the top left and then clicking on `details` of the Bugzilla plugin.

## Cherry Picking/Backporting

The Bugzilla automation can also assist in the backporting process by integrating with the cherry picking automation. To do an
automated cherry-pick, simply comment on GitHub `/cherrypick <<target-branch-name>>`. This will result in the PR being cherrypicked

> onto `<<target-branch-name>> after it merged and the new cherrypick-PR being assigned to you. After an automated cherry-pick was

Reviewer 2: ```suggestion
onto `<<target-branch-name>>` after it merged and the new cherrypick-PR being assigned to you. After an automated cherry-pick was
```

created for a Pull Request that referenced a Bugzilla bug, the automation will clone the original bug with the new target release
and retitle the PR to correctly reference that new Bug. The new bug will also be marked as dependent on the original bug.

> In order to get the backports to more then one past release right, you need to do it serially, i.E. cherrypick from the current release

Reviewer 2: ```suggestion
In order to get the backports to more then one past release right, you need to do it serially, _i.e._ cherry-pick from the current release
```
